### ISSUE TITLE
Potential security issue with BrowserWindow.loadURL

### ISSUE THREAD

**Issue by sancarn** [electron/electron#9656]
BrowserWindow.loadURL is a great way to implement websites into your own applications. For instance BrowserWindow.loadURL('http://google.com'), opens google in the current browser window, which is fantastic and I am really glad this feature exists. Another feature which I am glad exists is the ability for the BrowserWindow to execute NodeJS code.

With the ability to load URLs into the BrowserWindow things like custom WebBrowsers are one of the first thoughts that comes to mind. Or at least one that came to my mind when I first came across NodeJS. However you have to be very careful because since your browser can execute NodeJS code, any website you navigate to - or rather any website the user navigates to intentionally or otherwise - could have dire consequences.

As a test example I created the following website:
```
<!doctype html>
<html>
    <head>
        <title> Security Breach </title>
        <script>
            var fs=require('fs');
            var path=require('path');
            var filepath=path.join((require('os').homedir()),"Desktop","WebScriptedFile__NoPermsRequired.txt")
            fs.writeFileSync(filepath,"No permissions were required to write this file! ;)")
        </script>
    </head> 
    <body>
        Look at your desktop! ;)
    </body>
</html>
```

If loaded in a regular WebBrowser nothing interesting happens. However if you create a WebBrowser with Electron (and you don't implement any protection) the javascript will write a file to your desktop autonomously without your permission.

There is no denying that the fix for this security issue should probably come from the developer of the WebBrowser. However it would be nice if there was an option within Electron to sandbox (using an iframe) all **web resources** into loaded into a BrowserWindow. 

```BrowserWindow.sandboxAllWebResources = true/false```

Recently a friend and I wanted to do this dynamically (we're making a plugin system for an application we're making and want to enable this option by default) and since we can't modify electron at it's source (otherwise the application wouldn't stand the test of time) we need to override `require()` to override electron's browserwindow's loadURL function. Not a big deal, but it would be nice if such an option existed, and it would at least protect unsuspecting users from websites which exploit applications made by uninformed developers.

I still believe however that this rule should only apply to web content, and that local NodeJS code should execute as it does now. So correct me if I am wrong but this suggestion is different to the current existing sandbox and nodeIntegration options for the BrowserWindow. Ultimately in my opinion:

Local = Trustworthy
Online = Potentially untrustworthy

Thoughts?

Note:
An alternative solution would be an option in loadURL() to launch the URL in 'safe' mode or similar.

**Comment by javan** [comment]
These issues are covered pretty well in https://github.com/electron/electron/blob/master/docs/tutorial/security.md#checklist

**Comment by sancarn** [comment]
I hadn't ever seen this tutorial... That's a lot more potential issues than I had imagined even. It's good to know that these things will need to be implemented in our dynamic implementation!

**Comment by MarshallOfSound** [comment]
Thanks @javan, all this use case (and a lot of others) are covered in the linked [security checklist](https://github.com/electron/electron/blob/master/docs/tutorial/security.md#checklist)

The bit of information relevant to this is regarding `nodeIntegration` when displaying remote content.

> Disable the Node integration in all renderers that display remote content (setting nodeIntegration to false in webPreferences)

**Comment by enlight** [comment]
@tarruda implemented some sandbox functionality in PR #6919 that shipped in Electron 1.4.2, and you can find some discussion on sandboxed webviews in #9611. However, I'd like to emphasize this line by @tarruda in the aforementioned discussion:

> There's still a lot of work to be done before electron can be considered secure as chromium for running untrusted content.

**Comment by sancarn** [comment]
Thanks all for the feedback! 👍
